Bugzilla/Config.pm

```diff
@@ -11,375 +11,366 @@ use 5.10.1;
 use strict;
 use warnings;
 
-use base qw(Exporter);
+use base qw(Bugzilla::Object Exporter);
+
 use Bugzilla::Constants;
+use Bugzilla::Error;
 use Bugzilla::Hook;
-use Data::Dumper;
-use File::Temp;
+use Bugzilla::Logging;
+
 use Module::Runtime qw(require_module);
+use Safe;
+use Try::Tiny;
 
-# Don't export localvars by default - people should have to explicitly
-# ask for it, as a (probably futile) attempt to stop code using it
-# when it shouldn't
 %Bugzilla::Config::EXPORT_TAGS
-  = (admin => [qw(update_params SetParam write_params)],);
+  = (admin => [qw(SetParam write_params)],);
 Exporter::export_ok_tags('admin');
 
-# INITIALIZATION CODE
-# Perl throws a warning if we use bz_locations() directly after do.
-our %params;
-
-# Load in the param definitions
-sub _load_params {
-  my $panels = param_panels();
-  my %hook_panels;
-  foreach my $panel (keys %$panels) {
-    my $module = $panels->{$panel};
-    require_module($module);
-    my @new_param_list = $module->get_param_list();
-    $hook_panels{lc($panel)} = {params => \@new_param_list};
-  }
+###############################
+####    Initialization     ####
+###############################
 
-  # This hook is also called in editparams.cgi. This call here is required
-  # to make SetParam work.
-  Bugzilla::Hook::process('config_modify_panels', {panels => \%hook_panels});
+sub new {
+  my $invocant = shift;
+  my $class    = ref($invocant) || $invocant;
 
-  foreach my $panel (keys %hook_panels) {
-    foreach my $item (@{$hook_panels{$panel}->{params}}) {
-      $params{$item->{'name'}} = $item;
-    }
+  my $cache = Bugzilla->request_cache;
+  if ($cache->{params_obj}) {
+    return $cache->{params_obj};
   }
-}
 
-# END INIT CODE
+  my $self = {params => {}};
+  bless($self, $class);
```

Comment:
ref() returns the class that $invocant belongs to. In this case it would probably be `Bugzilla::Config`. Bless creates an pseudo object that Perl uses for OOP. So if I bless() a hash such as $self (normally what people name it, sort of like this in JS), then I can then use pointer notation ($object->do_something()) and any functions defined can be used as methods. The functions just assume the first param will be the $self reference.